### PR TITLE
Add callable filters

### PR DESCRIPTION
**Pull request by painor**
More type hints
Added multiple extensions
Add callable filters

Feel free to test/review before accepting

### CODE REVIEW

File: directory_downloader/directory_downloader.py [hunk 1]
```diff
@@ -47,24 +49,37 @@ async def fetch_file_links(self, url: str = None, filter: str = None, extension:
...
-                        await self.fetch_file_links(next_url, filter=filter, extension=extension)
+                        await self.fetch_file_links(next_url, filter=filter, extensions=extensions)
 
         return self.files_urls
 
-    async def _get_site_content(self, url: str):
+    def is_valid_link(self, name, filter: Union[str, callable] = None, extensions: List[str] = None) -> bool:
+        if filter:
+            if callable(filter):
+                return filter()
```
Comment (mahdinaija7) [inline directory_downloader/directory_downloader.py]: the function filter() must take a file_name as an argument

File: directory_downloader/directory_downloader.py [hunk 2]
```diff
@@ -47,24 +49,37 @@ async def fetch_file_links(self, url: str = None, filter: str = None, extension:
...
+            elif isinstance(filter, str):
+                return bool(re.match(filter, name))
+            else:
+                raise ValueError("filter needs to be either a callable or a string")
+
+        if extensions:
+            valid_extension = False
+            for extension in extensions:
+
+                if not name.endswith(extension):
```
Comment (mahdinaija7) [inline directory_downloader/directory_downloader.py]: it doesn't work like that
to check for a valid extension the name must ends with extension

File: directory_downloader/directory_downloader.py [hunk 3]
```diff
@@ -47,24 +49,37 @@ async def fetch_file_links(self, url: str = None, filter: str = None, extension:
...
+                        logging.info(f"{next_url}")
                         self.files_urls.add(next_url)
                     else:
-                        await self.fetch_file_links(next_url, filter=filter, extension=extension)
+                        await self.fetch_file_links(next_url, filter=filter, extensions=extensions)
 
         return self.files_urls
 
-    async def _get_site_content(self, url: str):
+    def is_valid_link(self, name, filter: Union[str, callable] = None, extensions: List[str] = None) -> bool:
```
Comment (mahdinaija7) [inline directory_downloader/directory_downloader.py]: what if i passed 2 kwargs (filter and extenstions ) together
the extensions condition will never be reached

**Review by mahdinaija7 — CHANGES_REQUESTED** [review]

File: directory_downloader/directory_downloader.py [hunk 4]
```diff
@@ -60,24 +57,37 @@ async def fetch_file_links(self, url: str = None, filter: Union[str, callable] =
...
-                return bool(re.match(filter, name))
-            else:
-                raise ValueError("filter needs to be either a callable or a string")
+        if filter and extensions:
+            if self._check_extension(extensions, name) and self._check_filter(filter, name):
+                return True
+            return False
+
+        if filter:
+            self._check_filter(filter, name)
```
Comment (painor) [inline directory_downloader/directory_downloader.py]: This doesn't return anything so the value will always be falsy (None)

File: directory_downloader/directory_downloader.py [hunk 5]
```diff
@@ -60,24 +57,37 @@ async def fetch_file_links(self, url: str = None, filter: Union[str, callable] =
...
 
         if extensions:
-            valid_extension = False
-            for extension in extensions:
+            self._check_extension(extensions, name)
```
Comment (painor) [inline directory_downloader/directory_downloader.py]: Check the previous comment

File: directory_downloader/directory_downloader.py [hunk 6]
```diff
@@ -130,3 +139,5 @@ async def download_files(self, urls: Set[str] = None, filter: Union[str, callabl
         connector = aiohttp.TCPConnector(limit=self.workers)
         async with aiohttp.ClientSession(connector=connector) as session:
             await self._start_downloads(session, filter, extensions)
+
+
```
Comment (painor) [inline directory_downloader/directory_downloader.py]: too many empty lines. one is enough

File: .gitignore [hunk 7]
```diff
@@ -126,4 +126,5 @@ venv.bak/
 dmypy.json
 
 # Pyre type checker
-.pyre/
\ No newline at end of file
+.pyre/
+.idea/workspace.xml
```
Comment (painor) [inline .gitignore]: add a new line

**Pull request merged**